core/src/main/scala/kafka/log/Log.scala

```diff
@@ -513,6 +513,11 @@ class Log(@volatile var dir: File,
           // we are taking the offsets we are given
           if (!appendInfo.offsetsMonotonic || appendInfo.firstOffset < nextOffsetMetadata.messageOffset)
             throw new IllegalArgumentException("Out of order offsets found in " + records.records.asScala.map(_.offset))
+
+          if (!records.hasCompatibleMagic(config.messageFormatVersion.messageFormatVersion)) {
+            trace(s"Down converting messages on replica fetching for $topicPartition")
+            validRecords = records.downConvert(config.messageFormatVersion.messageFormatVersion)
+          }
```

Comment:
Does it make sense to update the `leaderEpochCache` and then down-convert? cc @junrao